// @anysphere/priompt/dist/base.test.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
Object.defineProperty(exports, "__esModule", { value: true });
const vitest_1 = require("vitest");
const Priompt = __importStar(require("./index"));
const lib_1 = require("./lib");
const components_1 = require("./components");
const tokenizer_1 = require("./tokenizer");
(0, vitest_1.describe)("isolate", () => {
    function Isolate(props) {
        if (props.isolate) {
            return (Priompt.createElement(Priompt.Fragment, null,
                Priompt.createElement("isolate", { p: props.p, prel: props.prel, tokenLimit: props.tokenLimit }, props.children)));
        }
        else {
            return (Priompt.createElement(Priompt.Fragment, null,
                Priompt.createElement("scope", { p: props.p, prel: props.prel }, props.children)));
        }
    }
    function Test(props) {
        return (Priompt.createElement(Priompt.Fragment, null,
            "This is the start of the prompt.",
            Priompt.createElement(Isolate, { tokenLimit: 100, isolate: props.isolate }, Array.from({ length: 1000 }, (_, i) => (Priompt.createElement(Priompt.Fragment, null,
                Priompt.createElement("scope", { prel: -i - 2000 },
                    "This is an SHOULDBEINCLUDEDONLYIFISOLATED user message number",
                    " ",
                    i))))),
            Array.from({ length: 1000 }, (_, i) => (Priompt.createElement(Priompt.Fragment, null,
                Priompt.createElement("scope", { prel: -i - 1000 },
                    "This is user message number ",
                    i)))),
            Priompt.createElement(Isolate, { tokenLimit: 100, isolate: props.isolate }, Array.from({ length: 1000 }, (_, i) => (Priompt.createElement(Priompt.Fragment, null,
                Priompt.createElement("scope", { prel: -i },
                    i,
                    ",xl,x,,",
                    i > 100 ? "SHOULDBEINCLUDEDONLYIFNOTISOLATED" : "")))))));
    }
    (0, vitest_1.it)("should have isolate work", async () => {
        const renderedIsolated = await (0, lib_1.render)(Test({ isolate: true }), {
            tokenLimit: 1000,
            tokenizer: (0, tokenizer_1.getTokenizerByName_ONLY_FOR_OPENAI_TOKENIZERS)("cl100k_base"),
        });
        (0, vitest_1.expect)(renderedIsolated.tokenCount).toBeLessThanOrEqual(1000);
        (0, vitest_1.expect)((0, lib_1.isPlainPrompt)(renderedIsolated.prompt)).toBe(true);
        if (!(0, lib_1.isPlainPrompt)(renderedIsolated.prompt))
            return;
        (0, vitest_1.expect)(renderedIsolated.prompt.includes("SHOULDBEINCLUDEDONLYIFISOLATED")).toBe(true);
        (0, vitest_1.expect)(renderedIsolated.prompt.includes("SHOULDBEINCLUDEDONLYIFNOTISOLATED")).toBe(false);
        const renderedUnIsolated = await (0, lib_1.render)(Test({ isolate: false }), {
            tokenLimit: 1000,
            tokenizer: (0, tokenizer_1.getTokenizerByName_ONLY_FOR_OPENAI_TOKENIZERS)("cl100k_base"),
        });
        (0, vitest_1.expect)(renderedUnIsolated.tokenCount).toBeLessThanOrEqual(1000);
        (0, vitest_1.expect)((0, lib_1.isPlainPrompt)(renderedUnIsolated.prompt)).toBe(true);
        if (!(0, lib_1.isPlainPrompt)(renderedUnIsolated.prompt))
            return;
        (0, vitest_1.expect)(renderedUnIsolated.prompt.includes("SHOULDBEINCLUDEDONLYIFISOLATED")).toBe(false);
        (0, vitest_1.expect)(renderedUnIsolated.prompt.includes("SHOULDBEINCLUDEDONLYIFNOTISOLATED")).toBe(true);
    });
    function SimplePrompt(props) {
        return (Priompt.createElement(Priompt.Fragment, null,
            "This is the start of the p",
            props.breaktoken ? Priompt.createElement("breaktoken", null) : Priompt.createElement(Priompt.Fragment, null),
            "rompt. This is the second part of the prompt."));
    }
    (0, vitest_1.it)("promptToTokens should work", async () => {
        const donotbreak = await (0, lib_1.render)(SimplePrompt({ breaktoken: false }), {
            tokenLimit: 1000,
            tokenizer: (0, tokenizer_1.getTokenizerByName_ONLY_FOR_OPENAI_TOKENIZERS)("cl100k_base"),
        });
        const toTokens = await (0, lib_1.promptToTokens)(donotbreak.prompt, (0, tokenizer_1.getTokenizerByName_ONLY_FOR_OPENAI_TOKENIZERS)("cl100k_base"));
        (0, vitest_1.expect)(donotbreak.tokenCount).toBe(toTokens.length);
        (0, vitest_1.expect)(toTokens).toStrictEqual([
            2028, 374, 279, 1212, 315, 279, 10137, 13, 1115, 374, 279, 2132, 961, 315,
            279, 10137, 13,
        ]);
        const dobreak = await (0, lib_1.render)(SimplePrompt({ breaktoken: true }), {
            tokenLimit: 1000,
            tokenizer: (0, tokenizer_1.getTokenizerByName_ONLY_FOR_OPENAI_TOKENIZERS)("cl100k_base"),
        });
        (0, vitest_1.expect)(dobreak.tokenCount).toBe(donotbreak.tokenCount + 1);
        const toTokens2 = await (0, lib_1.promptToTokens)(dobreak.prompt, (0, tokenizer_1.getTokenizerByName_ONLY_FOR_OPENAI_TOKENIZERS)("cl100k_base"));
        (0, vitest_1.expect)(dobreak.tokenCount).toBe(toTokens2.length);
        (0, vitest_1.expect)(toTokens2).toStrictEqual([
            2028, 374, 279, 1212, 315, 279, 281, 15091, 13, 1115, 374, 279, 2132, 961,
            315, 279, 10137, 13,
        ]);
    });
    function SimpleMessagePrompt(props) {
        return (Priompt.createElement(Priompt.Fragment, null,
            Priompt.createElement(components_1.SystemMessage, null,
                "This is the start of the prompt.",
                Priompt.createElement("br", null),
                props.breaktoken ? Priompt.createElement("breaktoken", null) : Priompt.createElement(Priompt.Fragment, null),
                Priompt.createElement("br", null),
                "This is the second part of the prompt."),
            Priompt.createElement(components_1.UserMessage, null, "hi!")));
    }
    (0, vitest_1.it)("promptToTokens should work", async () => {
        const donotbreak = await (0, lib_1.render)(SimpleMessagePrompt({ breaktoken: false }), {
            tokenLimit: 1000,
            tokenizer: (0, tokenizer_1.getTokenizerByName_ONLY_FOR_OPENAI_TOKENIZERS)("cl100k_base"),
            lastMessageIsIncomplete: true,
        });
        const toTokens = await (0, lib_1.promptToTokens)(donotbreak.prompt, (0, tokenizer_1.getTokenizerByName_ONLY_FOR_OPENAI_TOKENIZERS)("cl100k_base"));
        (0, vitest_1.expect)(donotbreak.tokenCount).toBe(toTokens.length);
        (0, vitest_1.expect)(toTokens).toStrictEqual([
            100264, 9125, 100266, 2028, 374, 279, 1212, 315, 279, 10137, 382, 2028,
            374, 279, 2132, 961, 315, 279, 10137, 13, 100265, 100264, 882, 100266,
            6151, 0,
        ]);
        const dobreak = await (0, lib_1.render)(SimpleMessagePrompt({ breaktoken: true }), {
            tokenLimit: 1000,
            tokenizer: (0, tokenizer_1.getTokenizerByName_ONLY_FOR_OPENAI_TOKENIZERS)("cl100k_base"),
            lastMessageIsIncomplete: true,
        });
        (0, vitest_1.expect)(dobreak.tokenCount).toBe(donotbreak.tokenCount + 1);
        const toTokens2 = await (0, lib_1.promptToTokens)(dobreak.prompt, (0, tokenizer_1.getTokenizerByName_ONLY_FOR_OPENAI_TOKENIZERS)("cl100k_base"));
        (0, vitest_1.expect)(dobreak.tokenCount).toBe(toTokens2.length);
        (0, vitest_1.expect)(toTokens2).toStrictEqual([
            100264, 9125, 100266, 2028, 374, 279, 1212, 315, 279, 10137, 627, 198,
            2028, 374, 279, 2132, 961, 315, 279, 10137, 13, 100265, 100264, 882,
            100266, 6151, 0,
        ]);
    });
    function SpecialTokensPrompt() {
        return (Priompt.createElement(Priompt.Fragment, null,
            Priompt.createElement(components_1.SystemMessage, null, "<|im_start|>"),
            Priompt.createElement(components_1.UserMessage, null, "<|diff_marker|>"),
            Priompt.createElement(components_1.AssistantMessage, null, "<|endoftext|>")));
    }
    (0, vitest_1.it)("handle special tokens", async () => {
        const specialTokens = await (0, lib_1.render)(SpecialTokensPrompt(), {
            tokenLimit: 1000,
            tokenizer: (0, tokenizer_1.getTokenizerByName_ONLY_FOR_OPENAI_TOKENIZERS)("cl100k_base"),
            lastMessageIsIncomplete: true,
        });
        (0, vitest_1.expect)(specialTokens.tokenCount).toBeGreaterThanOrEqual(24);
        const toTokens = await (0, lib_1.promptToTokens)(specialTokens.prompt, (0, tokenizer_1.getTokenizerByName_ONLY_FOR_OPENAI_TOKENIZERS)("cl100k_base"));
        (0, vitest_1.expect)(specialTokens.tokenCount).toBe(toTokens.length);
        (0, vitest_1.expect)(toTokens).toStrictEqual([
            100264, 9125, 100266, 27, 91, 318, 5011, 91, 29, 100265, 100264, 882,
            100266, 27, 91, 13798, 27363, 91, 29, 100265, 100264, 78191, 100266, 27,
            91, 8862, 728, 428, 91, 29,
        ]);
    });
    (0, vitest_1.it)("handle all special tokens encoded", async () => {
        const specialTokens = await (0, lib_1.render)(SpecialTokensPrompt(), {
            tokenLimit: 1000,
            tokenizer: (0, tokenizer_1.getTokenizerByName_ONLY_FOR_OPENAI_TOKENIZERS)("cl100k_base_special_tokens"),
            lastMessageIsIncomplete: true,
        });
        (0, vitest_1.expect)(specialTokens.tokenCount).toBeGreaterThanOrEqual(24);
        const toTokens = await (0, lib_1.promptToTokens)(specialTokens.prompt, (0, tokenizer_1.getTokenizerByName_ONLY_FOR_OPENAI_TOKENIZERS)("cl100k_base_special_tokens"));
        (0, vitest_1.expect)(specialTokens.tokenCount).toBe(toTokens.length);
        (0, vitest_1.expect)(toTokens).toStrictEqual([
            100264, 9125, 100266, 100264, 100265, 100264, 882, 100266, 27, 91, 13798,
            27363, 91, 29, 100265, 100264, 78191, 100266, 27, 91, 8862, 728, 428, 91,
            29,
        ]);
    });
});
(0, vitest_1.describe)("config", () => {
    function TestConfig(props) {
        return (Priompt.createElement(Priompt.Fragment, null,
            "This is the start of the prompt.",
            Priompt.createElement("config", { stop: "\n" }),
            Priompt.createElement("config", { maxResponseTokens: "tokensReserved" })));
    }
    (0, vitest_1.it)("should have config work", async () => {
        const rendered = await (0, lib_1.render)(TestConfig({ numConfigs: 1 }), {
            tokenLimit: 1000,
            tokenizer: (0, tokenizer_1.getTokenizerByName_ONLY_FOR_OPENAI_TOKENIZERS)("cl100k_base"),
        });
        (0, vitest_1.expect)(rendered.tokenCount).toBeLessThanOrEqual(1000);
        (0, vitest_1.expect)((0, lib_1.isPlainPrompt)(rendered.prompt)).toBe(true);
        (0, vitest_1.expect)(rendered.config.stop).toBe("\n");
        (0, vitest_1.expect)(rendered.config.maxResponseTokens).toBe("tokensReserved");
    });
});
