// @anysphere/priompt/dist/components.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.ZFunction = exports.Function = exports.ZTools = exports.Tools = exports.ToolResultMessage = exports.FunctionMessage = exports.ImageComponent = exports.AssistantMessage = exports.UserMessage = exports.SystemMessage = void 0;
const Priompt = __importStar(require("./lib"));
const zod_to_json_schema_1 = __importDefault(require("zod-to-json-schema"));
function SystemMessage(props) {
    return {
        type: "chat",
        role: "system",
        name: props.name,
        to: props.to,
        children: props.children !== undefined
            ? Array.isArray(props.children)
                ? props.children.flat()
                : [props.children]
            : [],
    };
}
exports.SystemMessage = SystemMessage;
function UserMessage(props) {
    return {
        type: "chat",
        role: "user",
        name: props.name,
        to: props.to,
        children: props.children !== undefined
            ? Array.isArray(props.children)
                ? props.children.flat()
                : [props.children]
            : [],
    };
}
exports.UserMessage = UserMessage;
function AssistantMessage(props) {
    return {
        type: "chat",
        role: "assistant",
        functionCall: props.functionCall,
        toolCalls: props.toolCalls,
        to: props.to,
        children: props.children !== undefined
            ? Array.isArray(props.children)
                ? props.children.flat()
                : [props.children]
            : [],
    };
}
exports.AssistantMessage = AssistantMessage;
function ImageComponent(props) {
    return {
        type: "image",
        bytes: props.bytes,
        dimensions: props.dimensions,
        detail: props.detail,
    };
}
exports.ImageComponent = ImageComponent;
function FunctionMessage(props) {
    return {
        type: "chat",
        role: "function",
        name: props.name,
        to: props.to,
        children: props.children !== undefined
            ? Array.isArray(props.children)
                ? props.children.flat()
                : [props.children]
            : [],
    };
}
exports.FunctionMessage = FunctionMessage;
function ToolResultMessage(props) {
    return {
        type: "chat",
        role: "tool",
        name: props.name,
        to: props.to,
        children: props.children !== undefined
            ? Array.isArray(props.children)
                ? props.children.flat()
                : [props.children]
            : [],
    };
}
exports.ToolResultMessage = ToolResultMessage;
const populateOnStreamResponseObjectFromOnStream = (captureProps) => ({
    ...captureProps,
    onStreamResponseObject: async (stream) => {
        const messageStream = (async function* () {
            for await (const s of stream) {
                if (s.choices.length === 0)
                    continue;
                if (s.choices[0].delta === undefined)
                    continue;
                const message = s.choices[0].delta;
                yield message;
            }
        })();
        await captureProps.onStream(messageStream);
    },
});
// design choice: can only have 1 tools component per prompt, and cannot have any other onStream
function Tools(props) {
    return (Priompt.createElement(Priompt.Fragment, null,
        props.tools.map((tool) => (
        // eslint-disable-next-line @typescript-eslint/ban-ts-comment
        // @ts-ignore
        Priompt.createElement(Tool, { name: tool.name, description: tool.description, parameters: tool.parameters }))),
        populateOnStreamResponseObjectFromOnStream({
            type: "capture",
            onStream: async (stream) => {
                await props.onReturn((async function* () {
                    // index -> {name, args}
                    const toolCallsMap = new Map();
                    for await (const message of stream) {
                        if (message.content !== undefined) {
                            yield message.content;
                        }
                        // eslint-disable-next-line @typescript-eslint/ban-ts-comment
                        // @ts-ignore
                        const toolCalls = message.tool_calls;
                        if (!Array.isArray(toolCalls)) {
                            continue;
                        }
                        for (const toolCall of toolCalls) {
                            const { index, id } = toolCall;
                            let toolInfo = toolCallsMap.get(index);
                            if (toolInfo === undefined) {
                                toolInfo = { name: "", args: "", toolCallId: id };
                            }
                            if (toolCall.function.name !== undefined) {
                                toolInfo.name = toolCall.function.name;
                            }
                            if (toolCall.function.arguments !== undefined) {
                                toolInfo.args += toolCall.function.arguments;
                                const tool = props.tools.find((tool) => tool.name === (toolInfo === null || toolInfo === void 0 ? void 0 : toolInfo.name));
                                if (tool !== undefined &&
                                    tool.onFormatAndYield !== undefined) {
                                    // try parsing as JSON, if successful, yield the parsed JSON
                                    try {
                                        const parsedArgs = JSON.parse(toolInfo.args);
                                        yield tool.onFormatAndYield(toolInfo.args, toolInfo.toolCallId, toolInfo.name, index);
                                    }
                                    catch (error) {
                                        // do nothing
                                    }
                                }
                            }
                            toolCallsMap.set(index, toolInfo);
                        }
                    }
                    for (const [toolIndex, toolInfo] of toolCallsMap.entries()) {
                        if (toolInfo.name !== undefined &&
                            toolInfo.args !== undefined) {
                            const tool = props.tools.find((tool) => tool.name === toolInfo.name);
                            if (tool !== undefined && tool.onCall !== undefined) {
                                await tool.onCall(toolInfo.args, toolInfo.toolCallId, toolInfo.name, toolIndex);
                            }
                        }
                    }
                })());
            },
        })));
}
exports.Tools = Tools;
function ZTools(props) {
    return (
    // eslint-disable-next-line @typescript-eslint/ban-ts-comment
    // @ts-ignore
    Priompt.createElement(Tools, { tools: props.tools.map((tool) => ({
            name: tool.name,
            description: tool.description,
            parameters: zodToJsonSchema(tool.parameters),
            onCall: async (args, toolCallId, toolName, toolIndex) => {
                var _a;
                try {
                    const parsedArgs = tool.parameters.parse(JSON.parse(args));
                    await ((_a = tool.onCall) === null || _a === void 0 ? void 0 : _a.call(tool, parsedArgs, toolCallId, toolName, toolIndex));
                }
                catch (error) {
                    console.error(`Error parsing arguments for tool ${tool.name}:`, error);
                    if (tool.onParseError !== undefined) {
                        await tool.onParseError(error, args);
                    }
                    else {
                        throw error;
                    }
                }
            },
            onFormatAndYield: tool.onFormatAndYield
                ? async (args, toolCallId, toolName, toolIndex) => {
                    var _a, _b;
                    try {
                        const parsedArgs = tool.parameters.parse(JSON.parse(args));
                        return ((_b = (_a = tool.onFormatAndYield) === null || _a === void 0 ? void 0 : _a.call(tool, parsedArgs, toolCallId, toolName, toolIndex)) !== null && _b !== void 0 ? _b : args);
                    }
                    catch (error) {
                        console.error(`Error formatting arguments for tool ${tool.name}:`, error);
                        return args;
                    }
                }
                : undefined,
        })), onReturn: props.onReturn }));
}
exports.ZTools = ZTools;
function Tool(props) {
    return (Priompt.createElement(Priompt.Fragment, null, {
        type: "toolDefinition",
        tool: {
            type: "function",
            function: {
                name: props.name,
                description: props.description,
                parameters: props.parameters,
            },
        },
    }));
}
function Function(props) {
    if (!validFunctionName(props.name)) {
        throw new Error(`Invalid function name: ${props.name}. Function names must be between 1 and 64 characters long and may only contain a-z, A-Z, 0-9, and underscores.`);
    }
    // eslint-disable-next-line @typescript-eslint/ban-ts-comment
    // @ts-ignore
    return (Priompt.createElement(Priompt.Fragment, null,
        {
            type: "functionDefinition",
            name: props.name,
            description: props.description,
            parameters: props.parameters,
        },
        {
            type: "capture",
            onOutput: async (output) => {
                if (props.onCall !== undefined &&
                    output.function_call !== undefined &&
                    output.function_call.name === props.name &&
                    output.function_call.arguments !== undefined) {
                    await props.onCall(output.function_call.arguments);
                }
            },
        }));
}
exports.Function = Function;
// May contain a-z, A-Z, 0-9, and underscores, with a maximum length of 64 characters.
function validFunctionName(name) {
    return /^[a-zA-Z0-9_]{1,64}$/.test(name);
}
function ZFunction(props) {
    return (
    // eslint-disable-next-line @typescript-eslint/ban-ts-comment
    // @ts-ignore
    Priompt.createElement(Function, { name: props.name, description: props.description, parameters: zodToJsonSchema(props.parameters), onCall: async (rawArgs) => {
            if (props.onCall === undefined) {
                // do nothing
                return;
            }
            try {
                const args = props.parameters.parse(JSON.parse(rawArgs));
                await props.onCall(args);
            }
            catch (error) {
                if (props.onParseError !== undefined) {
                    await props.onParseError(error, rawArgs);
                }
                else {
                    throw error;
                }
            }
        } }));
}
exports.ZFunction = ZFunction;
function zodToJsonSchema(schema) {
    const fullSchema = (0, zod_to_json_schema_1.default)(schema, { $refStrategy: "none" });
    const { $schema, default: defaultVal, definitions, description, markdownDescription, ...rest } = fullSchema;
    // delete additionalProperties
    if ("additionalProperties" in rest) {
        // eslint-disable-next-line @typescript-eslint/ban-ts-comment
        // @ts-ignore
        delete rest.additionalProperties;
    }
    return rest;
}
