// @anysphere/priompt/dist/components.test.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
Object.defineProperty(exports, "__esModule", { value: true });
const vitest_1 = require("vitest");
const Priompt = __importStar(require("./index"));
const lib_1 = require("./lib");
const components_1 = require("./components");
const tokenizer_1 = require("./tokenizer");
(0, vitest_1.describe)("SystemMessage", () => {
    function TestSystemMessage(props) {
        return Priompt.createElement(components_1.SystemMessage, null, "hi this is a system message");
    }
    (0, vitest_1.it)("should create a chat message", async () => {
        const rendered = await (0, lib_1.render)(TestSystemMessage({}), {
            tokenLimit: 1000,
            tokenizer: (0, tokenizer_1.getTokenizerByName_ONLY_FOR_OPENAI_TOKENIZERS)("cl100k_base"),
        });
        (0, vitest_1.expect)((0, lib_1.isChatPrompt)(rendered.prompt)).toBe(true);
    });
    function TestSystemMessageWithName(props) {
        return (Priompt.createElement(Priompt.Fragment, null,
            Priompt.createElement(components_1.SystemMessage, { name: props.systemName }, "hi this is a system message"),
            Priompt.createElement(components_1.UserMessage, { name: props.userName }, "hi this is a user message"),
            Priompt.createElement(components_1.UserMessage, null, "hi this is a user message")));
    }
    (0, vitest_1.it)("should test the 'name' field of the SystemMessage", async () => {
        const rendered = await (0, lib_1.render)(TestSystemMessageWithName({ systemName: "TestName", userName: "carl" }), {
            tokenLimit: 1000,
            tokenizer: (0, tokenizer_1.getTokenizerByName_ONLY_FOR_OPENAI_TOKENIZERS)("cl100k_base"),
        });
        (0, vitest_1.expect)((0, lib_1.isChatPrompt)(rendered.prompt)).toBe(true);
        if (!(0, lib_1.isChatPrompt)(rendered.prompt))
            return;
        (0, vitest_1.expect)(rendered.prompt.messages[0].role === "system").toBe(true);
        if (rendered.prompt.messages[0].role !== "system")
            return;
        (0, vitest_1.expect)(rendered.prompt.messages[0].name).toBe("TestName");
        (0, vitest_1.expect)(rendered.prompt.messages[1].role === "user").toBe(true);
        if (rendered.prompt.messages[1].role !== "user")
            return;
        (0, vitest_1.expect)(rendered.prompt.messages[1].name).toBe("carl");
        (0, vitest_1.expect)(rendered.prompt.messages[2].role === "user").toBe(true);
        if (rendered.prompt.messages[2].role !== "user")
            return;
        (0, vitest_1.expect)(rendered.prompt.messages[2].name).toBeUndefined();
        const openAIPrompt = (0, lib_1.promptToOpenAIChatMessages)(rendered.prompt);
        (0, vitest_1.expect)(openAIPrompt[0].role === "system").toBe(true);
        if (openAIPrompt[0].role !== "system")
            return;
        (0, vitest_1.expect)(openAIPrompt[0].name).toBe("TestName");
        (0, vitest_1.expect)(openAIPrompt[1].role === "user").toBe(true);
        if (openAIPrompt[1].role !== "user")
            return;
        (0, vitest_1.expect)(openAIPrompt[1].name).toBe("carl");
        (0, vitest_1.expect)(openAIPrompt[2].role === "user").toBe(true);
        if (openAIPrompt[2].role !== "user")
            return;
        (0, vitest_1.expect)(openAIPrompt[2].name).toBeUndefined();
    });
});
(0, vitest_1.describe)("Function", () => {
    function TestFunction(props) {
        return (Priompt.createElement(Priompt.Fragment, null,
            Priompt.createElement(components_1.Function, { name: "echo", description: "Echo a message to the user.", parameters: {
                    type: "object",
                    properties: {
                        message: {
                            type: "string",
                            description: "The message to echo.",
                        },
                    },
                    required: ["message"],
                } }),
            Priompt.createElement(components_1.UserMessage, null, "say hi")));
    }
    (0, vitest_1.it)("should create a function message", async () => {
        const rendered = await (0, lib_1.render)(TestFunction({}), {
            tokenLimit: 1000,
            tokenizer: (0, tokenizer_1.getTokenizerByName_ONLY_FOR_OPENAI_TOKENIZERS)("cl100k_base"),
        });
        (0, vitest_1.expect)((0, lib_1.isChatPrompt)(rendered.prompt)).toBe(true);
        (0, vitest_1.expect)((0, lib_1.promptHasFunctions)(rendered.prompt)).toBe(true);
        if (!(0, lib_1.promptHasFunctions)(rendered.prompt))
            return;
        (0, vitest_1.expect)(rendered.prompt.functions).toEqual([
            {
                name: "echo",
                description: "Echo a message to the user.",
                parameters: {
                    type: "object",
                    properties: {
                        message: {
                            type: "string",
                            description: "The message to echo.",
                        },
                    },
                    required: ["message"],
                },
            },
        ]);
    });
});
(0, vitest_1.describe)("All kinds of messages", () => {
    function TestAllMessages(props) {
        return (Priompt.createElement(Priompt.Fragment, null,
            Priompt.createElement(components_1.Function, { name: "echo", description: "Echo a message to the user.", parameters: {
                    type: "object",
                    properties: {
                        message: {
                            type: "string",
                            description: "The message to echo.",
                        },
                    },
                    required: ["message"],
                } }),
            Priompt.createElement(components_1.SystemMessage, null, "System message"),
            Priompt.createElement(components_1.UserMessage, null, "User message"),
            Priompt.createElement(components_1.AssistantMessage, { functionCall: {
                    name: "echo",
                    arguments: '{"message": "this is a test echo"}',
                } }),
            Priompt.createElement(components_1.FunctionMessage, { name: "echo" }, "this is a test echo"),
            Priompt.createElement(components_1.AssistantMessage, { to: "python" }, "print(\"Hello world!\")"),
            Priompt.createElement(components_1.ToolResultMessage, { name: "python", to: "all" }, "Hello world!")));
    }
    (0, vitest_1.it)("should create all kinds of messages", async () => {
        const rendered = await (0, lib_1.render)(TestAllMessages({}), {
            tokenLimit: 1000,
            tokenizer: (0, tokenizer_1.getTokenizerByName_ONLY_FOR_OPENAI_TOKENIZERS)("cl100k_base"),
        });
        (0, vitest_1.expect)((0, lib_1.isChatPrompt)(rendered.prompt)).toBe(true);
        if (!(0, lib_1.isChatPrompt)(rendered.prompt))
            return;
        (0, vitest_1.expect)((0, lib_1.promptHasFunctions)(rendered.prompt)).toBe(true);
        if (!(0, lib_1.promptHasFunctions)(rendered.prompt))
            return;
        (0, vitest_1.expect)(rendered.prompt.functions).toEqual([
            {
                name: "echo",
                description: "Echo a message to the user.",
                parameters: {
                    type: "object",
                    properties: {
                        message: {
                            type: "string",
                            description: "The message to echo.",
                        },
                    },
                    required: ["message"],
                },
            },
        ]);
        (0, vitest_1.expect)(rendered.prompt.messages).toEqual([
            {
                role: "system",
                content: "System message",
            },
            {
                role: "user",
                content: "User message",
            },
            {
                role: "assistant",
                functionCall: {
                    name: "echo",
                    arguments: '{"message": "this is a test echo"}',
                },
            },
            {
                role: "function",
                name: "echo",
                content: "this is a test echo",
            },
            {
                role: "assistant",
                to: "python",
                content: 'print("Hello world!")',
            },
            {
                role: "tool",
                name: "python",
                to: "all",
                content: "Hello world!",
            },
        ]);
        const openaiMessage = (0, lib_1.promptToOpenAIChatMessages)(rendered.prompt);
        (0, vitest_1.expect)(openaiMessage.length).toBe(6);
        (0, vitest_1.expect)(openaiMessage[0].role).toBe("system");
        (0, vitest_1.expect)(openaiMessage[1].role).toBe("user");
        (0, vitest_1.expect)(openaiMessage[2].role).toBe("assistant");
        (0, vitest_1.expect)(openaiMessage[3].role).toBe("function");
        (0, vitest_1.expect)(openaiMessage[4].role).toBe("assistant");
        // the tool shall not be sent to openai! it's unsupported
        (0, vitest_1.expect)(openaiMessage[5].role).toBe("tool");
        // assert none of them contain "to"
        (0, vitest_1.expect)("to" in openaiMessage[4]).toBe(false);
        (0, vitest_1.expect)("to" in openaiMessage[5]).toBe(false);
    });
});
vitest_1.describe.skip("Images", () => {
    function TestImageMessage(props) {
        return (Priompt.createElement(Priompt.Fragment, null,
            Priompt.createElement(components_1.SystemMessage, null, "System message"),
            Priompt.createElement(components_1.UserMessage, null,
                Priompt.createElement(components_1.ImageComponent, { bytes: new Uint8Array([0, 0, 0, 0, 0]), dimensions: { width: 10, height: 10 }, detail: "auto" }),
                Priompt.createElement("br", null),
                "If the instructions mention this image, use it to help you write the code with the utmost precision and detail.",
                Priompt.createElement("br", null),
                "<instructions>",
                Priompt.createElement("br", null),
                "TEST, THIS IS A TEST,",
                Priompt.createElement("br", null),
                "</instructions>",
                Priompt.createElement("br", null))));
    }
    (0, vitest_1.it)("should create all kinds of messages", async () => {
        const rendered = await (0, lib_1.render)(TestImageMessage({}), {
            tokenLimit: 1000,
            tokenizer: (0, tokenizer_1.getTokenizerByName_ONLY_FOR_OPENAI_TOKENIZERS)("cl100k_base"),
        });
        (0, vitest_1.expect)((0, lib_1.isChatPrompt)(rendered.prompt)).toBe(true);
        if (!(0, lib_1.isChatPrompt)(rendered.prompt))
            return;
        // make sure the prompt string part is in the right order
        (0, vitest_1.expect)(rendered.prompt.messages[1].content).toBe("\n" +
            "If the instructions mention this image, use it to help you write the code with the utmost precision and detail.\n" +
            "<instructions>\n" +
            "TEST, THIS IS A TEST,\n" +
            "</instructions>\n");
    });
});
(0, vitest_1.describe)("Large prompts with many new lines and priority levels", () => {
    function TestLargePrompt(props) {
        return (Priompt.createElement(Priompt.Fragment, null,
            Priompt.createElement(components_1.SystemMessage, null, "You are a helpful assistant."),
            Priompt.createElement(components_1.UserMessage, null, Array(100)
                .fill(null)
                .map((_, i) => (Priompt.createElement("scope", { p: i * 10 },
                "This is line ",
                i + 1,
                " of the large prompt.",
                i % 10 === 0 && Priompt.createElement("br", null),
                i % 20 === 0 && Priompt.createElement("hr", null))))),
            Priompt.createElement(components_1.AssistantMessage, null, "Understood. How can I help you with this large prompt?"),
            Priompt.createElement(components_1.UserMessage, null,
                Priompt.createElement("scope", { p: 1000 }, "This is a high priority message that should always be included."),
                Priompt.createElement("scope", { p: 500 }, "This is a medium priority message that might be included."),
                Array(50)
                    .fill(null)
                    .map((_, i) => (Priompt.createElement("scope", { p: i * 10 },
                    "This is a message with priority ",
                    i * 10,
                    "."))))));
    }
    (0, vitest_1.it)("should render the large prompt correctly", async () => {
        var _a, _b;
        const rendered = await (0, lib_1.render)(TestLargePrompt({}), {
            tokenLimit: 1000,
            tokenizer: (0, tokenizer_1.getTokenizerByName_ONLY_FOR_OPENAI_TOKENIZERS)("cl100k_base"),
        });
        (0, vitest_1.expect)((0, lib_1.isChatPrompt)(rendered.prompt)).toBe(true);
        if (!(0, lib_1.isChatPrompt)(rendered.prompt))
            return;
        (0, vitest_1.expect)(rendered.prompt.messages.length).toBe(4);
        (0, vitest_1.expect)(rendered.prompt.messages[0].role).toBe("system");
        (0, vitest_1.expect)(rendered.prompt.messages[1].role).toBe("user");
        (0, vitest_1.expect)(rendered.prompt.messages[2].role).toBe("assistant");
        (0, vitest_1.expect)(rendered.prompt.messages[3].role).toBe("user");
        // Check that the high priority message is included
        (0, vitest_1.expect)(rendered.prompt.messages[3].content).toContain("This is a high priority message that should always be included.");
        // Check that some of the dynamic priority messages are included
        const dynamicPriorityCount = (_b = (_a = rendered.prompt.messages[3].content.match(/This is a message with priority/g)) === null || _a === void 0 ? void 0 : _a.length) !== null && _b !== void 0 ? _b : 0;
        (0, vitest_1.expect)(dynamicPriorityCount).toBeGreaterThan(0);
        (0, vitest_1.expect)(dynamicPriorityCount).toBeLessThan(50);
        // Check that the token count is within the limit
        (0, vitest_1.expect)(rendered.tokenCount).toBeLessThanOrEqual(2000);
    });
    (0, vitest_1.it)("should handle different token limits", async () => {
        var _a, _b;
        const tokenLimits = [500, 1000, 1500, 2000, 3000];
        for (const limit of tokenLimits) {
            const rendered = await (0, lib_1.render)(TestLargePrompt({}), {
                tokenLimit: limit,
                tokenizer: (0, tokenizer_1.getTokenizerByName_ONLY_FOR_OPENAI_TOKENIZERS)("cl100k_base"),
            });
            (0, vitest_1.expect)((0, lib_1.isChatPrompt)(rendered.prompt)).toBe(true);
            if (!(0, lib_1.isChatPrompt)(rendered.prompt))
                return;
            (0, vitest_1.expect)(rendered.tokenCount).toBeLessThanOrEqual(limit);
            // Check that the number of included messages increases with the token limit
            const userMessageContent = rendered.prompt.messages[3].content;
            const dynamicPriorityCount = (_b = (_a = userMessageContent.match(/This is a message with priority/g)) === null || _a === void 0 ? void 0 : _a.length) !== null && _b !== void 0 ? _b : 0;
            if (limit > 1000) {
                (0, vitest_1.expect)(dynamicPriorityCount).toBeGreaterThan(0);
            }
        }
    });
    function TestNestedPriorityLevels(props) {
        return (Priompt.createElement(components_1.UserMessage, null,
            Priompt.createElement("scope", { p: 1000 },
                "Top level high priority",
                Priompt.createElement("scope", { p: 500 },
                    "Nested medium priority",
                    Priompt.createElement("scope", { p: 100 },
                        "Deeply nested low priority",
                        Priompt.createElement("scope", { p: 50 }, "Very deeply nested very low priority")))),
            Array(10)
                .fill(null)
                .map((_, i) => (Priompt.createElement("scope", { p: (10 - i) * 100 },
                "Priority level ",
                (10 - i) * 100,
                Priompt.createElement("scope", { p: (10 - i) * 50 },
                    "Nested priority level ",
                    (10 - i) * 50))))));
    }
    (0, vitest_1.it)("should handle nested priority levels correctly", async () => {
        const rendered = await (0, lib_1.render)(TestNestedPriorityLevels({}), {
            tokenLimit: 70,
            tokenizer: (0, tokenizer_1.getTokenizerByName_ONLY_FOR_OPENAI_TOKENIZERS)("cl100k_base"),
        });
        (0, vitest_1.expect)((0, lib_1.isChatPrompt)(rendered.prompt)).toBe(true);
        if (!(0, lib_1.isChatPrompt)(rendered.prompt))
            return;
        const content = rendered.prompt.messages[0].content;
        (0, vitest_1.expect)(content).toContain("Top level high priority");
        (0, vitest_1.expect)(content).toContain("Nested medium priority");
        (0, vitest_1.expect)(content).toContain("Priority level 1000");
        (0, vitest_1.expect)(content).toContain("Priority level 900");
        // Lower priority items might be excluded due to token limit
        const lowerPriorityCount = (content.match(/Priority level [1-9]/g) || [])
            .length;
        (0, vitest_1.expect)(lowerPriorityCount).toBeLessThan(9);
        (0, vitest_1.expect)(rendered.tokenCount).toBeLessThanOrEqual(500);
    });
});
(0, vitest_1.describe)("Complex PromptElement Regression Tests", () => {
    const tokenizer = (0, tokenizer_1.getTokenizerByName_ONLY_FOR_OPENAI_TOKENIZERS)("cl100k_base");
    (0, vitest_1.it)("should render a complex PromptElement with nested elements", async () => {
        function ComplexPrompt(props) {
            return (Priompt.createElement(Priompt.Fragment, null,
                Priompt.createElement(components_1.SystemMessage, null,
                    "This is a complex system message with",
                    " ",
                    Priompt.createElement("scope", { p: 1000 }, "high priority content"),
                    " and",
                    " ",
                    Priompt.createElement("scope", { p: 500 }, "medium priority content"),
                    ".",
                    Priompt.createElement("br", null),
                    "New line in system message."),
                Priompt.createElement(components_1.UserMessage, null,
                    "User message with a ",
                    Priompt.createElement("breaktoken", null),
                    " break and ",
                    Priompt.createElement("br", null),
                    " line break.",
                    Priompt.createElement("scope", { p: 800 },
                        Priompt.createElement(components_1.ImageComponent, { bytes: new Uint8Array([0, 1, 2, 3]), dimensions: { width: 100, height: 100 }, detail: "auto" }),
                        "Image with some text")),
                Priompt.createElement(components_1.AssistantMessage, null,
                    "Assistant response with",
                    " ",
                    Priompt.createElement("scope", { prel: -100 }, "lower priority information"),
                    Priompt.createElement("breaktoken", null),
                    "After break token in assistant message"),
                Priompt.createElement(components_1.Function, { name: "testFunction", description: "A test function", parameters: {
                        type: "object",
                        properties: {
                            param1: { type: "string" },
                            param2: { type: "number" },
                        },
                        required: ["param1"],
                    } }),
                Priompt.createElement(components_1.FunctionMessage, { name: "testFunction" },
                    "Function result with nested scopes:",
                    Priompt.createElement("scope", { p: 700 },
                        Priompt.createElement("scope", { p: 750 }, "Highest priority"),
                        "Medium priority",
                        Priompt.createElement("br", null),
                        "New line in function message")),
                [1, 2, 3].map((num) => (Priompt.createElement(components_1.UserMessage, null,
                    "Message ",
                    num,
                    " in array",
                    num % 2 === 0 && (Priompt.createElement("scope", { p: 600 },
                        "Even numbered message",
                        Priompt.createElement("breaktoken", null),
                        "After break in even message")),
                    Priompt.createElement("br", null)))),
                Priompt.createElement("empty", { tokens: 5 })));
        }
        const rendered = await Priompt.render(ComplexPrompt({}), {
            tokenLimit: 1000,
            tokenizer,
            shouldBuildSourceMap: true,
        });
        // Remove 'durationMs' from rendered since its non deterministic for the snapshot
        delete rendered.durationMs;
        (0, vitest_1.expect)(rendered).toMatchSnapshot();
    });
    (0, vitest_1.it)("should render a PromptElement with multiple nested arrays and first elements", async () => {
        function NestedArraysPrompt(props) {
            return (Priompt.createElement(Priompt.Fragment, null,
                Priompt.createElement(components_1.SystemMessage, null,
                    "System message for nested arrays test",
                    Priompt.createElement("br", null),
                    "New line in system message"),
                [
                    [1, 2],
                    [3, 4],
                ].map((subArray, i) => (Priompt.createElement("scope", { p: 1000 - i * 100 }, subArray.map((num) => (Priompt.createElement(components_1.UserMessage, null,
                    "Nested array message ",
                    num,
                    Priompt.createElement("breaktoken", null),
                    "After break in nested message",
                    Priompt.createElement("first", null,
                        Priompt.createElement("scope", { p: 800 },
                            "First choice for ",
                            num),
                        Priompt.createElement("scope", { p: 700 },
                            "Second choice for ",
                            num),
                        Priompt.createElement("scope", { p: 600 },
                            "Third choice for ",
                            num,
                            Priompt.createElement("br", null),
                            "New line in third choice")))))))),
                Priompt.createElement(components_1.AssistantMessage, null,
                    Priompt.createElement("first", null,
                        Priompt.createElement("scope", { p: 900 },
                            "High priority assistant response",
                            Priompt.createElement("breaktoken", null),
                            "After break in high priority"),
                        Priompt.createElement("scope", { p: 800 }, "Medium priority assistant response"),
                        Priompt.createElement("scope", { p: 700 },
                            "Low priority assistant response",
                            Priompt.createElement("br", null),
                            "New line in low priority")))));
        }
        const rendered = await Priompt.render(NestedArraysPrompt({}), {
            tokenLimit: 1000,
            tokenizer,
            shouldBuildSourceMap: true,
        });
        // Remove 'durationMs' from rendered since its non deterministic for the snapshot
        delete rendered.durationMs;
        (0, vitest_1.expect)(rendered).toMatchSnapshot();
    });
    (0, vitest_1.it)("should render a PromptElement with isolate and capture elements", async () => {
        const capturedOutput = [];
        const capturedStream = [];
        const capturedStreamResponse = [];
        function IsolateAndCapturePrompt(props) {
            return (Priompt.createElement(Priompt.Fragment, null,
                Priompt.createElement(components_1.SystemMessage, null, "System message for isolate and capture test"),
                Priompt.createElement("isolate", { tokenLimit: 200 },
                    Priompt.createElement(components_1.UserMessage, null,
                        "This is an isolated user message with a token limit of 200",
                        Priompt.createElement("scope", { p: 900 }, "High priority isolated content"),
                        Priompt.createElement("scope", { p: 800 }, "Medium priority isolated content"),
                        Priompt.createElement("scope", { p: 700 }, "Low priority isolated content"))),
                Priompt.createElement("capture", { onOutput: async (output) => {
                        capturedOutput.push(output);
                    }, onStream: async (stream) => {
                        for await (const chunk of stream) {
                            capturedStream.push(JSON.stringify(chunk));
                        }
                    }, onStreamResponseObject: async (stream) => {
                        for await (const chunk of stream) {
                            capturedStreamResponse.push(chunk);
                        }
                    } }),
                Priompt.createElement(components_1.AssistantMessage, null,
                    "Assistant message after capture",
                    Priompt.createElement("scope", { p: 950 }, "Very high priority content"),
                    Priompt.createElement("scope", { p: 850 }, "High priority content"))));
        }
        const rendered = await Priompt.render(IsolateAndCapturePrompt({}), {
            tokenLimit: 1000,
            tokenizer,
            shouldBuildSourceMap: true,
        });
        // Simulate some output and streams to test the capture
        if (rendered.outputHandlers.length > 0) {
            await rendered.outputHandlers[0]({
                role: "assistant",
                content: "Test output",
            });
        }
        if (rendered.streamHandlers.length > 0) {
            const testStream = async function* () {
                yield { role: "assistant", content: "Stream " };
                yield { role: "assistant", content: "test " };
                yield { role: "assistant", content: "output" };
            };
            await rendered.streamHandlers[0](testStream());
        }
        if (rendered.streamResponseObjectHandlers.length > 0) {
            const testStreamResponse = async function* () {
                yield {
                    id: "1",
                    object: "chat.completion.chunk",
                    created: 1234567890,
                    model: "gpt-3.5-turbo",
                    choices: [
                        { index: 0, delta: { role: "assistant", content: "Response " } },
                    ],
                };
                yield {
                    id: "2",
                    object: "chat.completion.chunk",
                    created: 1234567891,
                    model: "gpt-3.5-turbo",
                    choices: [{ index: 0, delta: { role: "user", content: "object " } }],
                };
                yield {
                    id: "3",
                    object: "chat.completion.chunk",
                    created: 1234567892,
                    model: "gpt-3.5-turbo",
                    choices: [
                        { index: 0, delta: { role: "assistant", content: "stream" } },
                    ],
                };
            };
            await rendered.streamResponseObjectHandlers[0](testStreamResponse());
        }
        // Remove 'durationMs' from rendered since its non deterministic for the snapshot
        delete rendered.durationMs;
        // Include captured data in the snapshot
        const snapshotData = {
            rendered,
            capturedOutput,
            capturedStream,
            capturedStreamResponse,
        };
        (0, vitest_1.expect)(snapshotData).toMatchSnapshot();
    });
    (0, vitest_1.it)("should render a PromptElement with config and complex nested structures", async () => {
        function ComplexNestedPrompt(props) {
            return (Priompt.createElement(Priompt.Fragment, null,
                Priompt.createElement("config", { maxResponseTokens: 100, stop: ["STOP", "END"] }),
                Priompt.createElement(components_1.SystemMessage, null,
                    "System message for complex nested structures test",
                    Priompt.createElement("br", null),
                    "New line in system message"),
                Priompt.createElement(components_1.UserMessage, null,
                    "User message with nested structures",
                    Priompt.createElement("scope", { p: 900 },
                        "High priority scope",
                        Priompt.createElement("first", null,
                            Priompt.createElement("scope", { p: 950 },
                                "First choice in high priority",
                                Priompt.createElement("breaktoken", null),
                                "After break in first choice"),
                            Priompt.createElement("scope", { p: 940 }, "Second choice in high priority")),
                        Priompt.createElement("isolate", { tokenLimit: 150 },
                            Priompt.createElement(components_1.UserMessage, null,
                                "Nested isolated user message",
                                Priompt.createElement("br", null),
                                "New line in isolated message"),
                            Priompt.createElement(components_1.AssistantMessage, null,
                                "Nested isolated assistant message",
                                Priompt.createElement("breaktoken", null),
                                "After break in isolated assistant message"))),
                    Priompt.createElement("scope", { p: 800 },
                        "Medium priority scope",
                        Priompt.createElement("breaktoken", null),
                        "After break token",
                        Priompt.createElement(components_1.ImageComponent, { bytes: new Uint8Array([4, 5, 6, 7]), dimensions: { width: 200, height: 200 }, detail: "high" }))),
                Priompt.createElement(components_1.AssistantMessage, { functionCall: {
                        name: "testFunction",
                        arguments: JSON.stringify({ param1: "test", param2: 42 }),
                    } },
                    "Assistant message with function call",
                    Priompt.createElement("br", null),
                    "New line in assistant message"),
                Priompt.createElement(components_1.FunctionMessage, { name: "testFunction" },
                    "Function result for complex nested test",
                    Priompt.createElement("breaktoken", null),
                    "After break in function message"),
                Priompt.createElement("empty", { tokens: 10 })));
        }
        const rendered = await Priompt.render(ComplexNestedPrompt({}), {
            tokenLimit: 1500,
            tokenizer,
            shouldBuildSourceMap: true,
        });
        // Remove 'durationMs' from rendered since its non deterministic for the snapshot
        delete rendered.durationMs;
        (0, vitest_1.expect)(rendered).toMatchSnapshot();
    });
    (0, vitest_1.it)("should render a complex PromptElement with nested elements and arrays", async () => {
        function ComplexPrompt(props) {
            const priorityLevels = [1000, 800, 600, 400, 200, 100];
            const messageTypes = ["user", "assistant", "system"];
            return (Priompt.createElement(Priompt.Fragment, null,
                Priompt.createElement(components_1.SystemMessage, null, "This is a complex system message with nested elements and arrays."),
                priorityLevels.map((priority, index) => (Priompt.createElement("scope", { p: priority },
                    Priompt.createElement(components_1.UserMessage, null,
                        "Priority level ",
                        priority,
                        " message",
                        index % 2 === 0 && Priompt.createElement("breaktoken", null),
                        [...Array(20)].map((_, i) => (Priompt.createElement("scope", { p: priority - i * 50 },
                            "Nested scope ",
                            i + 1,
                            " in priority ",
                            priority,
                            Priompt.createElement("br", null)))))))),
                Priompt.createElement(components_1.AssistantMessage, null,
                    "Assistant message with nested elements",
                    Priompt.createElement("first", null, messageTypes.map((type, index) => (Priompt.createElement("scope", { p: 900 - index * 100 },
                        type === "user" && (Priompt.createElement(components_1.UserMessage, null, "Nested user message")),
                        type === "assistant" && (Priompt.createElement(components_1.AssistantMessage, null, "Nested assistant message")),
                        type === "system" && (Priompt.createElement(components_1.SystemMessage, null, "Nested system message"))))))),
                [...Array(3)].map((_, i) => (Priompt.createElement(components_1.Function, { name: `testFunction${i + 1}`, description: `Test function ${i + 1}`, parameters: {
                        type: "object",
                        properties: {
                            param1: { type: "string" },
                            param2: { type: "number" },
                        },
                        required: ["param1"],
                    } }))),
                Priompt.createElement("isolate", { tokenLimit: 300 }, [...Array(4)].map((_, i) => (Priompt.createElement("scope", { p: 1000 - i * 100 },
                    Priompt.createElement(components_1.UserMessage, null,
                        "Isolated message ",
                        i + 1,
                        Priompt.createElement(components_1.ImageComponent, { bytes: new Uint8Array([i, i + 1, i + 2, i + 3]), dimensions: { width: 100 * (i + 1), height: 100 * (i + 1) }, detail: i % 2 === 0 ? "auto" : "high" })))))),
                [...Array(3)].map((_, i) => (Priompt.createElement(components_1.FunctionMessage, { name: `testFunction${i + 1}` },
                    "Function ",
                    i + 1,
                    " result with nested scopes:",
                    [...Array(3)].map((_, j) => (Priompt.createElement("scope", { p: 700 - j * 50 },
                        "Nested scope ",
                        j + 1,
                        " in function ",
                        i + 1,
                        j === 1 && Priompt.createElement("breaktoken", null))))))),
                Priompt.createElement("empty", { tokens: 10 })));
        }
        const rendered = await Priompt.render(ComplexPrompt({}), {
            tokenLimit: 5000,
            tokenizer,
            shouldBuildSourceMap: true,
        });
        delete rendered.durationMs;
        (0, vitest_1.expect)(rendered).toMatchSnapshot();
    });
    (0, vitest_1.it)("should render a PromptElement with multiple nested arrays and conditional elements", async () => {
        function NestedArraysPrompt(props) {
            const generateNestedArray = (rows, cols) => {
                return Array.from({ length: rows }, (_, i) => Array.from({ length: cols }, (_, j) => i * cols + j + 1));
            };
            const nestedArray = generateNestedArray(6, 6);
            return (Priompt.createElement(Priompt.Fragment, null,
                Priompt.createElement(components_1.SystemMessage, null, "System message for nested arrays and conditional elements test"),
                nestedArray.map((subArray, i) => (Priompt.createElement("scope", { p: 1000 - i * 100 }, subArray.map((num) => (Priompt.createElement(components_1.UserMessage, null,
                    "Nested array message ",
                    num,
                    Priompt.createElement("isolate", { tokenLimit: 500 },
                        Priompt.createElement("breaktoken", null),
                        Priompt.createElement("first", null, [...Array(10)].map((_, j) => (Priompt.createElement("scope", { p: 800 - j * 50 },
                            "Choice ",
                            j + 1,
                            " for ",
                            num))))))))))),
                Priompt.createElement(components_1.AssistantMessage, null,
                    Priompt.createElement("first", null, [...Array(10)].map((_, i) => (Priompt.createElement("scope", { p: 900 - i * 50 },
                        "Priority ",
                        900 - i * 50,
                        " assistant response",
                        Priompt.createElement("br", null),
                        Priompt.createElement("br", null),
                        i % 2 === 0 && Priompt.createElement("breaktoken", null),
                        i % 3 === 0 && (Priompt.createElement(components_1.ImageComponent, { bytes: new Uint8Array([i, i + 1, i + 2, i + 3]), dimensions: { width: 50 * (i + 1), height: 50 * (i + 1) }, detail: "auto" })))))))));
        }
        const rendered = await Priompt.render(NestedArraysPrompt({}), {
            tokenLimit: 5000,
            tokenizer,
            shouldBuildSourceMap: true,
        });
        delete rendered.durationMs;
        (0, vitest_1.expect)(rendered).toMatchSnapshot();
    });
});
