// @anysphere/priompt/dist/preview.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.register = exports.PreviewManager = exports.defaultYamlLoad = exports.defaultYamlDump = exports.dumpProps = exports.configFromSynchronousPrompt = exports.configFromPrompt = void 0;
const lib_1 = require("./lib");
const fs = __importStar(require("fs"));
const path = __importStar(require("path"));
const yaml = __importStar(require("js-yaml"));
const tokenizer_1 = require("./tokenizer");
const outputCatcher_ai_1 = require("./outputCatcher.ai");
function getProjectRoot() {
    if (process.env.PRIOMPT_PREVIEW_BASE_ABSOLUTE_PATH !== undefined) {
        return process.env.PRIOMPT_PREVIEW_BASE_ABSOLUTE_PATH;
    }
    if (process.env.PRIOMPT_PREVIEW_BASE_RELATIVE_PATH !== undefined) {
        return path.join(process.cwd(), process.env.PRIOMPT_PREVIEW_BASE_RELATIVE_PATH);
    }
    // just do cwd / priompt for now
    return process.cwd();
}
function configFromPrompt(prompt) {
    if (prompt.config !== undefined) {
        return prompt.config;
    }
    return {
        id: prompt.name,
        prompt,
    };
}
exports.configFromPrompt = configFromPrompt;
function configFromSynchronousPrompt(prompt, options) {
    if (prompt.config !== undefined) {
        return prompt.config;
    }
    return {
        id: prompt.name,
        prompt,
    };
}
exports.configFromSynchronousPrompt = configFromSynchronousPrompt;
function dumpProps(config, props) {
    let hasNoDump = false;
    for (const key in props) {
        if (key.startsWith('DO_NOT_DUMP')) {
            hasNoDump = true;
        }
    }
    let objectToDump = props;
    if (hasNoDump) {
        objectToDump = {};
        for (const key in props) {
            if (!key.startsWith('DO_NOT_DUMP')) {
                // eslint-disable-next-line @typescript-eslint/ban-ts-comment
                // @ts-ignore
                objectToDump[key] = props[key];
            }
        }
    }
    const dump = config.dump
        ? config.dump(objectToDump)
        : defaultYamlDump(objectToDump);
    return dump;
}
exports.dumpProps = dumpProps;
function defaultYamlDump(props) {
    // JSON is valid YAML, so yaml loads of this data will work fine.
    // the YAML dump is synchronous and relatively slow, and JSON stringify is much faster.
    return JSON.stringify(props, null, 2);
}
exports.defaultYamlDump = defaultYamlDump;
function defaultYamlLoad(dump) {
    // This has to load both yaml data (prior to 9-17-24) and json data (post 9-17-24)
    // json is valid yaml so this will work, but yaml is not valid json, so a json load would not work.
    return yaml.load(dump);
}
exports.defaultYamlLoad = defaultYamlLoad;
class PreviewManagerImpl {
    get shouldDump() {
        return this._shouldDump;
    }
    set shouldDump(value) {
        this._shouldDump = value;
    }
    getConfig(promptId) {
        return this.previews[promptId];
    }
    getPreviews() {
        return Object.keys(this.previews).reduce((acc, promptId) => {
            const promptPath = path.join(getProjectRoot(), 'priompt', promptId);
            const dumpsPath = path.join(promptPath, 'dumps');
            // if they don't exist, make the dirs
            if (!fs.existsSync(promptPath)) {
                fs.mkdirSync(promptPath, { recursive: true });
            }
            if (!fs.existsSync(dumpsPath)) {
                fs.mkdirSync(dumpsPath, { recursive: true });
            }
            const propsIds = fs.readdirSync(dumpsPath).filter((f) => f.endsWith('.yaml')).map((f) => f.replace('.yaml', ''));
            const savedIds = fs.readdirSync(promptPath).filter((f) => f.endsWith('.yaml')).map((f) => f.replace('.yaml', ''));
            return {
                ...acc,
                [promptId]: {
                    dumps: propsIds,
                    saved: savedIds,
                }
            };
        }, {});
    }
    async getPrompt(query) {
        let element = exports.PreviewManager.getElement(query.promptId, query.propsId);
        if (element instanceof Promise) {
            element = await element;
        }
        const rendered = await (0, lib_1.render)(element, { tokenizer: (0, tokenizer_1.getTokenizerByName_ONLY_FOR_OPENAI_TOKENIZERS)(query.tokenizer), tokenLimit: query.tokenLimit, shouldBuildSourceMap: query.shouldBuildSourceMap });
        return rendered;
    }
    async getPromptOutput(query) {
        const outputCatcher = (0, outputCatcher_ai_1.NewOutputCatcher)();
        let element = exports.PreviewManager.getElement(query.promptId, query.propsId, outputCatcher);
        if (element instanceof Promise) {
            element = await element;
        }
        const rendered = await (0, lib_1.render)(element, { tokenizer: (0, tokenizer_1.getTokenizerByName_ONLY_FOR_OPENAI_TOKENIZERS)(query.tokenizer), tokenLimit: query.tokenLimit, shouldBuildSourceMap: query.shouldBuildSourceMap });
        if (!query.stream) {
            // call all of them and wait all of them in parallel
            await Promise.all(rendered.outputHandlers.map((handler) => handler(Array.isArray(query.completion) ? query.completion[0] : query.completion)));
            // now return the first output
            const firstOutput = outputCatcher.getOutput();
            return firstOutput;
        }
        else {
            await Promise.all(rendered.streamHandlers.map((handler) => handler((async function* () {
                for (const completion of Array.isArray(query.completion) ? query.completion : [query.completion]) {
                    yield completion;
                }
            })())));
            // now return the first output
            const firstOutput = outputCatcher.getOutput();
            // let's just put it in an array
            const a = [];
            // eslint-disable-next-line @typescript-eslint/no-explicit-any
            for await (const x of firstOutput) {
                a.push(x);
            }
            return a;
        }
    }
    async getPromptFromRemote(query) {
        let element = this.getRemoteElement(query.promptId, query.promptDump);
        if (element instanceof Promise) {
            element = await element;
        }
        return this.getPromptFromRemoteElement(query, element);
    }
    async getPropsFromRemote(query) {
        const promptId = query.promptId;
        const promptDump = query.promptDump;
        const config = this.previews[promptId];
        const baseProps = this.hydrate(config, promptDump);
        return baseProps;
    }
    getPromptFunctionFromRemote(query) {
        const promptId = query.promptId;
        const config = this.previews[promptId];
        return config.prompt;
    }
    async getPromptFromRemoteElement(query, element) {
        const rendered = await (0, lib_1.render)(element, { tokenizer: (0, tokenizer_1.getTokenizerByName_ONLY_FOR_OPENAI_TOKENIZERS)(query.tokenizer), tokenLimit: query.tokenLimit });
        return rendered;
    }
    getRemoteElement(promptId, promptDump) {
        const config = this.previews[promptId];
        const baseProps = this.hydrate(config, promptDump);
        const element = config.prompt(baseProps);
        return element;
    }
    getElement(promptId, propsId, outputCatcher) {
        if (promptId === 'liveModePromptId') {
            if (this.lastLiveModeData === null) {
                throw new Error('live mode prompt not found');
            }
            return this.lastLiveModeData.promptElement;
        }
        if (!Object.keys(this.previews).includes(promptId)) {
            throw new Error(`preview promptId ${promptId} not registered`);
        }
        const config = this.previews[promptId];
        const baseProps = this.hydrate(config, this.getDump(promptId, propsId));
        let realProps = baseProps;
        if (outputCatcher !== undefined) {
            const captureProps = {
                onReturn: (x) => outputCatcher.onOutput(x),
            };
            realProps = {
                // eslint-disable-next-line @typescript-eslint/no-explicit-any
                ...baseProps,
                // eslint-disable-next-line @typescript-eslint/no-explicit-any
                ...captureProps,
            };
        }
        return config.prompt(realProps);
    }
    registerConfig(config) {
        var _a;
        if (Object.keys(this.previews).includes(config.id)) {
            // sort of sketchy, but may be fine if we're in esm hmr land...
            // we just overwrite
            if (process.env.ALLOW_PROMPT_REREGISTRATION === "true") {
                console.warn(`preview id ${config.id} already registered`);
            }
            else {
                throw new Error(`preview id ${config.id} already registered`);
            }
        }
        if (((_a = config.prompt.config) === null || _a === void 0 ? void 0 : _a.id) !== undefined && config.prompt.config.id !== config.id && process.env.NODE_ENV === 'development') {
            throw new Error(`Prompt id ${config.prompt.config.id} does not match config id ${config.id}. Prompts and configs need to be in a 1-to-1 mapping, because otherwise the serialization/deserialization of props becomes unclear and may cause bugs for you. (If you want to register an alias name for a prompt, just wrap the prompt in another prompt and register that one.)`);
        }
        config.prompt.config = config;
        this.previews[config.id] = config;
    }
    register(prompt) {
        const config = configFromPrompt(prompt);
        this.registerConfig(config);
    }
    registerProtoPrompt(prompt, protoProps) {
        const config = configFromPrompt(prompt);
        config.dump = (props) => props.toJsonString();
        config.hydrate = (dump) => {
            try {
                return protoProps.fromJsonString(dump);
            }
            catch (e) {
                // we try parsing as yaml! for backwards compatibility!
                const jsonS = JSON.stringify(defaultYamlLoad(dump));
                return protoProps.fromJsonString(jsonS);
            }
        };
        this.registerConfig(config);
    }
    configFromSynchronousPrompt(prompt) {
        return configFromSynchronousPrompt(prompt);
    }
    hydrate(config, dump) {
        if (config.hydrate) {
            return config.hydrate(dump);
        }
        const yamlData = defaultYamlLoad(dump);
        const props = yamlData;
        return props;
    }
    maybeDump(prompt, props) {
        if (!this.shouldDump) {
            return;
        }
        const config = configFromPrompt(prompt);
        this.dump(config, props);
    }
    dump(config, props) {
        const dump = dumpProps(config, props);
        const priomptPath = path.join(getProjectRoot(), 'priompt', config.id);
        const dumpsPath = path.join(priomptPath, 'dumps');
        if (!fs.existsSync(priomptPath)) {
            fs.mkdirSync(priomptPath, { recursive: true });
            // in this case, we want to write the file to the promptId path as well, as example01.yaml
            // this makes it easier for other users to see what the prompt is supposed to look like
            const filePath = path.join(priomptPath, `example01.yaml`);
            fs.writeFileSync(filePath, dump);
        }
        if (!fs.existsSync(dumpsPath)) {
            fs.mkdirSync(dumpsPath, { recursive: true });
        }
        // if there are more than 2000 files, delete the oldest 1000 of them
        try {
            const files = fs.readdirSync(dumpsPath);
            if (files.length > 2000) {
                const sortedFiles = files.sort((a, b) => a.localeCompare(b));
                for (let i = 0; i < 1000; i++) {
                    fs.unlinkSync(path.join(dumpsPath, sortedFiles[i]));
                }
            }
        }
        catch (e) {
            console.warn({ error: e }, "failed to remove old priompt dumps");
        }
        const propsId = new Date().toISOString().replace(/[:.]/g, '-'); // Human-readable propsId with date and time
        const filePath = path.join(dumpsPath, `${propsId}.yaml`); // Changed file extension to .yaml
        fs.writeFileSync(filePath, dump);
    }
    constructor() {
        this._shouldDump = process.env.NODE_ENV === 'development';
        this.previews = {};
        this.lastLiveModeData = null;
        this.resolveLastLiveModeOutputPromise = () => { };
        this.resolveLiveModeResult = () => { };
        this.lastLiveModeOutputPromise = new Promise((resolve) => {
            this.resolveLastLiveModeOutputPromise = resolve;
        });
        this.liveModeResultPromise = new Promise((resolve) => {
            this.resolveLiveModeResult = resolve;
        });
    }
    async getLiveModePromptCompletion(promptElement, options) {
        const liveModeData = {
            liveModeId: randomString(),
            promptElement,
        };
        this.lastLiveModeData = liveModeData;
        this.resolveLastLiveModeOutputPromise();
        this.lastLiveModeOutputPromise = new Promise((resolve) => {
            this.resolveLastLiveModeOutputPromise = resolve;
        });
        const result = await this.liveModeResultPromise;
        const output = {
            'id': liveModeData.liveModeId,
            'object': 'text_completion',
            'created': Date.now(),
            'model': options.model,
            'choices': [
                {
                    'message': {
                        'role': 'assistant',
                        'content': result,
                    }
                }
            ]
        };
        return output;
    }
    async *streamLiveModePromptCompletion(promptElement, options) {
        const output = await this.getLiveModePromptCompletion(promptElement, options);
        output.choices[0].delta = output.choices[0].message;
        yield output;
    }
    async liveMode(query, abortSignal) {
        while (true) {
            const result = await Promise.race([
                this.lastLiveModeOutputPromise,
                new Promise((_, reject) => {
                    if (abortSignal) {
                        abortSignal.addEventListener('abort', () => reject(new Error('Aborted')));
                    }
                }),
            ]);
            if (result instanceof Error) {
                throw result;
            }
            if (this.lastLiveModeData === null) {
                continue;
            }
            if (this.lastLiveModeData.liveModeId === query.alreadySeenLiveModeId) {
                continue;
            }
            return this.lastLiveModeData;
        }
    }
    liveModeResult(query) {
        this.resolveLiveModeResult(query.output);
        this.liveModeResultPromise = new Promise((resolve) => {
            this.resolveLiveModeResult = resolve;
        });
    }
    getDump(promptId, propsId) {
        const priomptPath = path.join(getProjectRoot(), 'priompt', promptId);
        const dumpsPath = path.join(priomptPath, 'dumps');
        const filePathInPromptId = path.join(priomptPath, `${propsId}.yaml`);
        const filePathInDumps = path.join(dumpsPath, `${propsId}.yaml`);
        if (fs.existsSync(filePathInPromptId)) {
            return fs.readFileSync(filePathInPromptId, 'utf-8');
        }
        else if (fs.existsSync(filePathInDumps)) {
            return fs.readFileSync(filePathInDumps, 'utf-8');
        }
        else {
            throw new Error(`No dump found for promptId ${promptId} and propsId ${propsId}`);
        }
    }
}
// GLOBALS FTW. i love globals.
exports.PreviewManager = new PreviewManagerImpl();
// Decorator
function register() {
    // eslint-disable-next-line @typescript-eslint/no-explicit-any
    function registerPrompt(target, propertyKey, descriptor) {
        if (descriptor.value === undefined) {
            throw new Error(`@registerPrompt can only be used on methods, not ${target.constructor.name}.${propertyKey}`);
        }
        else {
            exports.PreviewManager.register(descriptor.value);
        }
    }
    return registerPrompt;
}
exports.register = register;
function randomString() {
    let s = '';
    for (let i = 0; i < 10; i++) {
        s += Math.floor(Math.random() * 10);
    }
    return s;
}
