// @anysphere/priompt/dist/sourcemap.test.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
Object.defineProperty(exports, "__esModule", { value: true });
const vitest_1 = require("vitest");
const Priompt = __importStar(require("./index"));
const lib_1 = require("./lib");
const components_1 = require("./components");
const tokenizer_1 = require("./tokenizer");
const validateSourceMap = (absoluteSourceMap, text) => {
    // Leaves which are raw strings are annotated with the string field
    // If we have their absolute range they should be an exact match
    const getLeaves = (sourceMap) => {
        const children = sourceMap.children || [];
        if (children.length > 0) {
            return children.flatMap((child) => getLeaves(child));
        }
        return [sourceMap];
    };
    const leaves = getLeaves(absoluteSourceMap);
    const leavesWithString = leaves.filter((leaf) => leaf.string !== undefined);
    const incorrectLeaves = leavesWithString.filter((leaf) => {
        const expectedString = leaf.string;
        const actualString = text.substring(leaf.start, leaf.end);
        return expectedString !== actualString;
    });
    if (incorrectLeaves.length > 0) {
        console.log("Failed to validate source map, incorrect leaves are", incorrectLeaves.sort((a, b) => a.start - b.start));
        return false;
    }
    return true;
};
const absolutifySourceMap = (sourceMap, offset = 0) => {
    var _a;
    const newOffset = sourceMap.start + offset;
    const children = ((_a = sourceMap.children) === null || _a === void 0 ? void 0 : _a.map((child) => absolutifySourceMap(child, newOffset))) ||
        [];
    return {
        ...sourceMap,
        start: newOffset,
        end: sourceMap.end + offset,
        children,
    };
};
(0, vitest_1.describe)("sourcemap", () => {
    function TestPromptTrivial(props) {
        return (Priompt.createElement(Priompt.Fragment, null,
            Priompt.createElement(components_1.SystemMessage, null, props.message),
            Priompt.createElement(components_1.UserMessage, null,
                "Testing sourcemap!",
                Priompt.createElement("br", null),
                "abcdef")));
    }
    function TestPromptEmojiAndJapanese() {
        return (Priompt.createElement(Priompt.Fragment, null,
            Priompt.createElement(components_1.SystemMessage, null, "\uD83E\uDEE8"),
            Priompt.createElement(components_1.UserMessage, null,
                Priompt.createElement("scope", { name: "lemon" }, "\uD83C\uDF4B"),
                Priompt.createElement("scope", { name: "japanese" }, "\u3053\u308C\u306F\u30EC\u30E2\u30F3"))));
    }
    function TestSourceMapMoreComplex(props) {
        const lines = props.message.split("\n");
        return (Priompt.createElement(Priompt.Fragment, null,
            Priompt.createElement(components_1.SystemMessage, null, "The System Message"),
            Priompt.createElement(components_1.UserMessage, { p: 1000 },
                Priompt.createElement("scope", { name: "the first line" },
                    "This is the first line",
                    Priompt.createElement("br", null),
                    Priompt.createElement("scope", { name: "buffer line" }, lines[0]),
                    Priompt.createElement("br", null)),
                Priompt.createElement("scope", { prel: -10, name: "lines" }, lines.map((line, i) => (Priompt.createElement("scope", { prel: -i }, line))))),
            Priompt.createElement(components_1.UserMessage, null,
                Priompt.createElement("scope", { name: "code" }, "const rendered = await render(TestPromptSplit(props.message)) tokenLimit: 100, tokenizer: getTokenizerByName(\"cl100k_base\"), buildSourceMap: true,"),
                Priompt.createElement("scope", { name: "checks", prel: -10 }, "const promptString = Priompt.promptToString_VULNERABLE_TO_PROMPT_INJECTION(rendered.prompt, getTokenizerByName(\"cl100k_base\")) const sourceMap = rendered.sourceMap"),
                Priompt.createElement("scope", { name: "expects" },
                    "expect(sourceMap).toBeDefined() expect(sourceMap?.start).toBe(0)",
                    Priompt.createElement("scope", { name: "lets put the mesage lines here again" }, props.message.split("\n").map((line, i) => (Priompt.createElement("scope", { name: `line ${i}` }, line)))),
                    "expect(sourceMap?.end).toBe(promp",
                    Priompt.createElement("scope", { name: "suspicious inline scope" }, "tString.length)"),
                    "expect(",
                    Priompt.createElement("scope", { name: "validate", prel: -5 },
                        "validateSourceMap(",
                        Priompt.createElement("scope", { name: "absolutify", prel: -4 }, "absolutifySourceMap(sourceMap!, 0)"),
                        ",",
                        Priompt.createElement("scope", { name: "promptString", prel: -3 }, "promptString"),
                        ")"),
                    ").toBe(true)"))));
    }
    (0, vitest_1.it)("should generate simple sourcemap correctly", async () => {
        var _a, _b;
        const rendered = await (0, lib_1.render)(TestPromptTrivial({ message: "System message for sourcemap test." }), {
            tokenLimit: 1000,
            tokenizer: (0, tokenizer_1.getTokenizerByName_ONLY_FOR_OPENAI_TOKENIZERS)("cl100k_base"),
            shouldBuildSourceMap: true,
        });
        (0, vitest_1.expect)(rendered.sourceMap).toBeDefined();
        const promptString = Priompt.promptToString_VULNERABLE_TO_PROMPT_INJECTION(rendered.prompt, (0, tokenizer_1.getTokenizerByName_ONLY_FOR_OPENAI_TOKENIZERS)("cl100k_base"));
        (0, vitest_1.expect)((_a = rendered.sourceMap) === null || _a === void 0 ? void 0 : _a.start).toBe(0);
        (0, vitest_1.expect)((_b = rendered.sourceMap) === null || _b === void 0 ? void 0 : _b.end).toBe(promptString.length);
        if (rendered.sourceMap === undefined) {
            throw new Error("bad");
        }
        (0, vitest_1.expect)(validateSourceMap(absolutifySourceMap(rendered.sourceMap, 0), promptString)).toBe(true);
    });
    (0, vitest_1.it)("should work with emoji and japanese", async () => {
        var _a, _b;
        const rendered = await (0, lib_1.render)(TestPromptEmojiAndJapanese(), {
            tokenLimit: 1000,
            tokenizer: (0, tokenizer_1.getTokenizerByName_ONLY_FOR_OPENAI_TOKENIZERS)("cl100k_base"),
            shouldBuildSourceMap: true,
        });
        (0, vitest_1.expect)(rendered.sourceMap).toBeDefined();
        const promptString = Priompt.promptToString_VULNERABLE_TO_PROMPT_INJECTION(rendered.prompt, (0, tokenizer_1.getTokenizerByName_ONLY_FOR_OPENAI_TOKENIZERS)("cl100k_base"));
        (0, vitest_1.expect)((_a = rendered.sourceMap) === null || _a === void 0 ? void 0 : _a.start).toBe(0);
        (0, vitest_1.expect)((_b = rendered.sourceMap) === null || _b === void 0 ? void 0 : _b.end).toBe(promptString.length);
        if (rendered.sourceMap === undefined) {
            throw new Error("bad");
        }
        (0, vitest_1.expect)(validateSourceMap(absolutifySourceMap(rendered.sourceMap, 0), promptString)).toBe(true);
    });
    (0, vitest_1.it)("should generate sourcemap correctly", async () => {
        const message = `one lever that we haven’t really touched is
    creating features that depend substantially
    on the user investing time
    in configuring them to work well
    eg if a user could spend a day configuring an agent
    that would consistently make them 50% more productive, it’d be worth it
    or a big company spending a month to
    structure their codebase in such a way that many more things than before can be ai automated.
    The polish we’re missing on cmd+k is making it rly fast and snappy for the 1-2 line use case
    but it should be able to handle 100s of lines of code in a reasonable timeframe
    `;
        const rendered = await (0, lib_1.render)(TestSourceMapMoreComplex({ message }), {
            tokenLimit: 300,
            tokenizer: (0, tokenizer_1.getTokenizerByName_ONLY_FOR_OPENAI_TOKENIZERS)("cl100k_base"),
            shouldBuildSourceMap: true,
        });
        const promptString = Priompt.promptToString_VULNERABLE_TO_PROMPT_INJECTION(rendered.prompt, (0, tokenizer_1.getTokenizerByName_ONLY_FOR_OPENAI_TOKENIZERS)("cl100k_base"));
        const sourceMap = rendered.sourceMap;
        (0, vitest_1.expect)(sourceMap).toBeDefined();
        (0, vitest_1.expect)(sourceMap === null || sourceMap === void 0 ? void 0 : sourceMap.start).toBe(0);
        (0, vitest_1.expect)(sourceMap === null || sourceMap === void 0 ? void 0 : sourceMap.end).toBe(promptString.length);
        if (sourceMap === undefined) {
            throw new Error("bad");
        }
        (0, vitest_1.expect)(validateSourceMap(absolutifySourceMap(sourceMap, 0), promptString)).toBe(true);
    });
});
